proxy/request.go

```diff
@@ -235,3 +252,25 @@ func (r *request) handleErrorResult(raw *frame.RawFrame) (retried bool) {
 
 	return retried
 }
+
+func (r *request) isBatchIdempotent(batch *partialBatch) (idempotent bool, err error) {
+	for _, queryOrId := range batch.queryOrIds {
+		switch q := queryOrId.(type) {
+		case string:
+			if idempotent, err = parser.IsQueryIdempotent(q); !idempotent {
+				return idempotent, err
+			}
+		case []byte:
+			var hash [16]byte
+			copy(hash[:], q)
+			if maybeIdempotent, ok := r.client.preparedIdempotence.Load(hash); !ok {
+				return false, errors.New("unable to determine if prepared statement in batch is idempotent")
```

Comment:
You're right, there's something off here. 

> It might be worth implementing a mechanism to return UNPREPARED when a cache miss occurs on the proxy (or re-preparing on the fly like the drivers do).

It does do this, but maybe not for batches.

Reply:
Ah, never mind. We are caching prepared statements and re-preparing automatically. This was added as part of this PR:
https://github.com/datastax/cql-proxy/pull/28

The relevant code is here: https://github.com/datastax/cql-proxy/blob/v0.0.4/proxycore/clientconn.go#L293-L328

Reply:
This error case is possible, but is highly unlikely to happen.

Reply:
So it's safe to assume that if this particular look up fails then we have to handle it (assuming `idempotent == false`) without repreparing because it was probably caused by something that would not get fixed by repreparing?

Reply:
I see a better way to do this using the existing prepared cache instead of having yet another map.

Reply:
> So it's safe to assume that if this particular look up fails then we have to handle it (by assuming idempotent == false) without repreparing because it was probably caused by something that would not get fixed by repreparing?

It could happen if the proxy receives a prepared ID that it's never seen before, but I think that would be an application bug.

> I see a better way to do this using the existing prepared cache instead of having yet another map. Trying to think of a way to make it clean-ish.